dll/win32/uxtheme/buffer.c

```diff
@@ -19,6 +19,28 @@
  */
 
 #include "uxthemep.h"
+#include <wine/heap.h>
```

Comment:
I do think the actual code-changes should also be guarded by something like
`#if DLL_EXPORT_VERSION >= _WIN32_WINNT_VISTA`

Reply:
If this was synced via the script? no.
If this was synced via copy and paste? That’s fine since they won’t be used unless that is set anyway.

@Splitwirez which did you do?

Reply:
???
Ofc they will be used, the implementation of e.g. BeginBufferedPaint(), ... changed, even for NT5 build.

Reply:
@JoachimHenze Those aren't even exported in NT5.2, are they?

Reply:
buffer.c is compiled even for NT5.2 builds:
![image](https://github.com/user-attachments/assets/7d29b4ab-58ab-4136-af48-983a8adaf4d9)
and its object file does contain BeginBufferedPaint():
![image](https://github.com/user-attachments/assets/53723c68-a921-4bed-befd-71d48add1143)

Reply:
Huh...that's unexpected. I could have sworn those APIs were only introduced in Windows Vista. In fact, the official Microsoft docs even support my understanding:
- [BeginBufferedPaint](https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-beginbufferedpaint) ([archived version from Jan 7, 2022](https://web.archive.org/web/20220107031650/https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-beginbufferedpaint))
- [DrawThemeBackground](https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-drawthemebackground) ([archived version from Jan 6, 2022](https://web.archive.org/web/20220106223315/https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-drawthemebackground))
Granted, Microsoft's documentation can be very...*weird* about whether or not to acknowledge sufficiently old versions of Windows, so I'm still not quite sure whether I can trust my memory on this front.

Furthermore, they aren't exported at all at present, [even in upstream master](https://github.com/reactos/reactos/blob/master/dll/win32/uxtheme/uxtheme.spec)...unless I'm missing something here? Are those files *not* the be-all-end-all of what gets exported?

To clarify, I'm *not* trying to say that you're wrong - on the contrary...I have no idea what I'm doing, but I want to learn. I want to understand why this matters.

### In short:
Why did I think `BeginBufferedPaint` and friends don't exist in Windows Server 2003? Am I stupid?

Furthermore, if  they aren't exported in Server 2003, then...why exactly do the changes matter?

Reply:
> I could have sworn those APIs were only introduced in Windows Vista

Buffered paint, yes but not DrawThemeBackground (it even says "Minimum supported server	Windows Server 2003 [desktop apps only]" on the current live page).

Both DrawThemeBackground and DrawThemeBackgroundEx (ordinal only?) are exported on 2003 and probably XP.

Using archive.org on pages that new is pointless, you can't use learn.microsoft.com nor docs.microsoft.com, you need to use the old msdn.microsoft.com and even then, you can't trust MSDN.

Reply:
@whindsaks `DrawThemeBackground` 100% exists in XP, yes. Sorry, I messed up how I phrased that - I'd only meant to mention the `DrawThemeBackground` docs to show that docs from the time period in question *do* acknowledge Server 2003 (at least sometimes?), but my head wasn't screwed on quite right so I forgot to actually say that lmao

The archive.org versions are included in case Microsoft decides to start weirdly specifically erasing history even further - in short, those links are there for tomorrow, not for today.

Reply:
From my pov what everybody here wrote doesn't contradict anything.
What I write here I did not double-check, but from my understanding
- DrawThemeBackground and DrawThemeBackgroundEx (ordinal only?) are exported on 2003 and probably XP.
- but BeginBufferedPaint is not from the MS tools

What I actually want is: If you change implementation of e.g. BeginBufferedPaint(), you are free to do that, but ideally those changes should be guarded by some ifdef, so we will bake them only into the binary when the target is defined to be NT6+.
A 5.2 build of our uxtheme should ideally not have them, neither implemented nor exported.
If that exclusion isn't the case yet, then it should be added.

They were inside the binaries already before you touched the implementation of BeginBufferedPaint(), and yes, that is and was an existing flaw most likely. But since they are now changed and are beginning to be more than just stubs, the exclusion should be added now.

Reply:
Using BeginBufferedPaint inside uxtheme is harmless IMHO. We have a time-machine of sorts here, we know BeginBufferedPaint is coming but we don't know exactly when MS started using it internally, we only know it was exported and documented starting with Vista.

Reply:
This **could** be correct @whindsaks , I cannot prove your claim to be wrong. It is possible that MS used it internally already before they started exporting it. I don't know how we could verify or falsify that hypothesis though, therefore my brain defaults to `try-to-be-careful-what-we-do-add`-mode.

Reply:
Yes but the point is, if uxtheme internally needs similar functionality (even just for [bitmap handling without drawing](https://devblogs.microsoft.com/oldnewthing/20110520-00/?p=10613)), it is better to model it after an API we know is coming.

Reply:
I do agree with that part, but we should take extra-care to not break anything regarding backwards compatibility and should not bloat configurations where specific stuff might not be needed (yet). But sure, **in the long term** moving into the same direction as MS did is definitely better than cooking our own soup entirely. That doesn't invalidate the intent of my discussion-thread though.